### PR TITLE
Export Box with `virsh dumpxml` with Additional Box Root Disks

### LINKED ISSUE #602: Multiple disks in Vagrant box / metadata.json
Hello.

Is it possible to add the "Multiple disks in Vagrant box" into the vagrant-libvirt? For example the official Virtualbox Vagrant box for Ubuntu 16.04 contains two disks:

```
$ wget -q https://atlas.hashicorp.com/ubuntu/boxes/xenial64/versions/20160521.0.0/providers/virtualbox.box -O - | tar tvf -
-rw-r--r-- root/root     11013 2016-05-21 16:36 box.ovf
-rw-r--r-- root/root       652 2016-05-21 16:36 Vagrantfile
-rw-r--r-- root/root        31 2016-05-21 16:36 metadata.json
-rw-r--r-- root/root       314 2016-05-21 16:36 ubuntu-xenial-16.04-cloudimg.mf
-rw-r--r-- root/root 290942976 2016-05-21 16:36 ubuntu-xenial-16.04-cloudimg.vmdk
-rw-r--r-- root/root     72192 2016-05-21 16:36 ubuntu-xenial-16.04-cloudimg-configdrive.vmdk
```

Is it possible to extend the vagrant-libvirt plugin to be able to support more than one disk in box file? (probably some metadata.json enhancement will be needed)
### Steps to reproduce
1. See the output of

```
$ wget -q https://atlas.hashicorp.com/ubuntu/boxes/xenial64/versions/20160521.0.0/providers/virtualbox.box -O - | tar tvf -
-rw-r--r-- root/root     11013 2016-05-21 16:36 box.ovf
-rw-r--r-- root/root       652 2016-05-21 16:36 Vagrantfile
-rw-r--r-- root/root        31 2016-05-21 16:36 metadata.json
-rw-r--r-- root/root       314 2016-05-21 16:36 ubuntu-xenial-16.04-cloudimg.mf
-rw-r--r-- root/root 290942976 2016-05-21 16:36 ubuntu-xenial-16.04-cloudimg.vmdk
-rw-r--r-- root/root     72192 2016-05-21 16:36 ubuntu-xenial-16.04-cloudimg-configdrive.vmdk
```
### Expected behaviour

Vagrant libvirt boxes (metadata.json) should support more than one disk.
### Actual behaviour

The Vagrant libvirt boxes has always one disk.
### System configuration

**OS/Distro version:**:

```
$ head -1 /etc/fedora-release
Fedora release 23 (Twenty Three)
```

**Libvirt version:**

```
$ rpm -qa  libvirt-daemon
libvirt-daemon-1.2.18.3-1.fc23.x86_64
```

**Output of `vagrant version; vagrant plugin list`:**

```
$ vagrant version; vagrant plugin list
Installed Version: 1.8.1
Latest Version: 1.8.1

You're running an up-to-date version of Vagrant!
vagrant-aws (0.7.0)
vagrant-cachier (1.2.1, system)
vagrant-hostmanager (1.8.1, system)
vagrant-libvirt (0.0.33, system)
```

**Output of `VAGRANT_LOG=debug vagrant ... --provider=libvirt`**
<none>

**A Vagrantfile to reproduce the issue:**
<none>

**Are you using upstream vagrant package or your distros package?**
Distro

### LINKED ISSUE #1013: File system not fully synced on `vagrant package`
### Steps to reproduce

1. Customize a ppc64el guest vagrant-libvirt box with a provisioning script.
2. `vagrant package`
3. Import the resulting box.
4. `vagrant up` the resulting box.

### Expected behaviour

All customizations specified in the provisioning script, such as systemd modifications, appear in the exported box.

### Actual behaviour

* The upstream box is not shutdown correctly during the `vagrant package` procedure.
* Many modifications provisioned on the upstream box, fail to appear on the downstream box.

For context, my shutdown command configured originally in the root base box with Packer, is `echo 'vagrant' | sudo -S halt -p`. This command appears to work normally when triggered by `vagrant halt`. It's just the `vagrant package` command interfacing with vagrant-libvirt that somehow fails to shutdown the box and thereby sync the file system to virtual disk persistence.

I can confirm that all the customizations are present in the parent box (either by `vagrant ssh` or VNC). However, neither the ensure-host-key shell script nor the ensure-host-key service are present on the child box (`vagrant ssh` broken, hence working on these scripts; use VNC instead).

### System configuration
**OS/Distro version:** Ubuntu 19.04 Disco Dingo amd64

**Libvirt version:** 5.0.0-1ubuntu2.3

**Output of `vagrant version; vagrant plugin list`:**

```console
$ vagrant version; vagrant plugin list
2.2.4
vagrant-libvirt (0.0.45, global)
vagrant-rsync-back (0.0.1, global)
```

bootstrap.sh:

```sh
#!/bin/sh
sudo apt-get update &&
    sudo apt-get install -y clang &&
    sudo apt-get autoclean -y &&
    sudo apt-get clean -y &&
    sudo rm -rf /var/lib/apt/lists/* \
        /var/cache/apt/pkgcache.bin \
        /var/cache/apt/srcpkgcache.bin


sudo cp /tmp/ensure-host-keys /usr/sbin/ensure-host-keys &&
    sudo chmod u+x /usr/sbin/ensure-host-keys &&
    sudo cp /tmp/ensure-host-keys.service /lib/systemd/system &&
    sudo systemctl enable ensure-host-keys
```

ensure-host-keys:

```sh
#!/bin/sh
test \
    -f /etc/ssh/ssh_host_rsa_key \
    -a -f /etc/ssh/ssh_host_ecdsa_key \
    -a -f /etc/ssh/ssh_host_ed25519_key ||
    dpkg-reconfigure openssh-server
```

ensure-host-keys.service:

```
[Unit]
Description=Prepare sshd host keys
Before=ssh.service

[Service]
Type=oneshot
RemainAfterExit=true
StandardOutput=journal
ExecStart=/usr/sbin/ensure-host-keys

[Install]
WantedBy=multi-user.target
```

**A Vagrantfile to reproduce the issue:**

```ruby
Vagrant.configure("2") do |config|
    config.vm.box = "mcandre/debian-ppc64el"
    config.vm.box_version = "0.0.1"

    # Prepare for vagrant package
    config.ssh.insert_key = false
    config.vm.synced_folder ".", "/vagrant", type: "rsync", disabled: true

    config.vm.provision "file", source: "ensure-host-keys", destination: "/tmp/ensure-host-keys"
    config.vm.provision "file", source: "ensure-host-keys.service", destination: "/tmp/ensure-host-keys.service"

    config.vm.provision "shell", path: "bootstrap.sh"
end
```

**Are you using upstream vagrant package or your distros package?**

Distro

### Workaround

Forcibly write `sudo sync` at the end of my provisioning scripts.

### LINKED ISSUE #1019: vagrant package fails with qemu >= 2.10
Vagrant package fails if the disk to package has a backing image and  the installed qemu version is greater than 2.10 while the virtual machine is still running.

The command fails with the following output:
```
qemu-img: Could not open '/var/lib/libvirt/images/[...].img': Failed to get shared "write" lock
Is another process using the image?
```

This openstack-nova issue lead me to the fix of the issue:
https://bugzilla.redhat.com/show_bug.cgi?id=1654804#c4

**To fix this issue you need to change the following line:**
Change
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/3f30606021444ac922c398e1737e09be4fe8fcfd/lib/vagrant-libvirt/action/package_domain.rb#L30

to:
```
backing = `qemu-img info --force-share "#{root_disk.path}" | grep 'backing file:' | cut -d ':' -f2`.chomp
```

Referring to the man page of `qemu-img` the `--force-share` option does the following:
```
If specified, qemu-img will open the image in shared mode, allowing other QEMU processes to open it in write mode. For example, this can be used to get the image information (with 'info' subcommand) when the image is used by a running guest.  Note that this could produce inconsistent results because of concurrent metadata changes, etc. This option is only allowed when opening images in read-only mode.
```

### Steps to reproduce
1. Install vagrant-libvirt on qemu >= 2.10 environment
2. Run and customize any vagrant box via libvirt provider
[Do not shutdown the virtual machine!]
3. Package the customized box via `vagrant package --output [...].box`

### Expected behaviour
The box should be packaged including copying and rebasing backing image

### Actual behaviour
The package command fails with:
```
qemu-img: Could not open '/var/lib/libvirt/images/[...].img': Failed to get shared "write" lock
Is another process using the image?
```

### System configuration
**OS/Distro version:**: RHEL 8

**Libvirt version:** 4.5.0

**qemu-img version:** 2.12.0 (qemu-kvm-2.12.0-63.module+el8+2833+c7d6d092)

**Output of `vagrant version; vagrant plugin list`:** 
Installed Version: 2.2.4
vagrant-libvirt (0.0.45, global)

**Are you using upstream vagrant package or your distros package?**
Upstream

### LINKED ISSUE #1256: Export/Import Box as Original VM with `virsh dumpxml` and `virsh define`?
A Standard Virtualbox Vagrant box `metadata.json` coming with following simple data structure, then reading all VM definition from bundled `box.ovf`:
```
{
  "provider": "virtualbox"
}
```

Currently our box `metadata.json` coming with following custom format, e.g.
```
{
  "format": "qcow2",
  "provider": "libvirt",
  "virtual_size": 128
}
```

For #602 and #1143 we are discussing add multiple disk support in our Vagrant box format, by [adding additional key/value into `metadata.json`](https://github.com/vagrant-libvirt/vagrant-libvirt/pull/1143/files#diff-515fec02588d6d41f3df9bce232b39a18d0352959c7b51f68fa459f55ef97a52R18-R30), e.g.
```
{
  "provider": "libvirt",
  "format": "qcow2",
  "disks": [
    {
      "virtual_size": 2
    },
    {
      "path": "disk2.qcow2",
      "virtual_size": 10
    }
  ]
}
```

With this direction, if we going to support additional NIC/USB/COM/etc device into vagrant-libvirt, we will keep bumping the custom structure into `metadata.json`, which is not scalable ;-(

What if we [export our VM with `virsh dumpxml`, then import it with `virsh define`](https://serverfault.com/a/707564/628984)? This come with a lot of benefits:

- Keeping `metadata.json` clean and tidy
- Support all original VM definition, e.g. `box.xml` (similar as Virtualbox's `box.ovf` format and functionality)

 
### Steps to reproduce
During `vagrant package`:
1. Export VM definition with `virsh dumpxml`
2. Search the exported `box.xml` for the disk image path and copy to local directory for packaging
3. Package both `metadata.json`, `Vagrantfile`, `box.xml`, and `*.qcow2` as our Vagrant Box archive

During `vagrant up`:
1. Restore original VM definition with `virsh define`
2. Search the `box.xml` for disk image filename, and copy to target directory, and update the VM definition with new disk image absolute path
3. Regenerate a new `<uuid>` for the imported VM
4. Remove some parameters that may conflict, e.g. `<cpu>`, `<interface>`, etc
4. Overwrite other else VM definition as per our existing logic

### Expected behaviour
Extracted box archive should coming with following files, e.g.:
```
box.xml
metadata.json
alvistack-VAGRANTSLASH-ubuntu-20.04_vagrant_box_image_20210425.0.0_0.img
vagrant-libvirt-vagrant-libvirt_default-vdb.qcow2
Vagrantfile
```

Where `metadata.json` simplified as: 
```
{
  "provider": "libvirt"
}
```

And `box.xml` in format that export by `virsh xmldump` and could import by `virsh define`, e.g.
```
<domain type='kvm' id='1'>
  <name>vagrant-libvirt-vagrant-libvirt_default</name>
  <uuid>778d5ab8-997e-4b11-82a6-cb6eb595cb32</uuid>
  <memory unit='KiB'>524288</memory>
  <currentMemory unit='KiB'>524288</currentMemory>
  <vcpu placement='static'>1</vcpu>
  <resource>
    <partition>/machine</partition>
  </resource>
  <os>
    <type arch='x86_64' machine='pc-i440fx-hirsute'>hvm</type>
    <boot dev='hd'/>
  </os>
  <features>
    <acpi/>
    <apic/>
    <pae/>
  </features>
  <cpu mode='custom' match='exact' check='full'>
    <model fallback='forbid'>Icelake-Client</model>
    <vendor>Intel</vendor>
    <feature policy='require' name='ss'/>
    <feature policy='require' name='vmx'/>
    <feature policy='require' name='pdcm'/>
    <feature policy='require' name='hypervisor'/>
    <feature policy='require' name='tsc_adjust'/>
    <feature policy='require' name='avx512f'/>
    <feature policy='require' name='avx512dq'/>
    <feature policy='require' name='avx512ifma'/>
    <feature policy='require' name='clflushopt'/>
    <feature policy='require' name='avx512cd'/>
    <feature policy='require' name='sha-ni'/>
    <feature policy='require' name='avx512bw'/>
    <feature policy='require' name='avx512vl'/>
    <feature policy='require' name='rdpid'/>
    <feature policy='require' name='fsrm'/>
    <feature policy='require' name='md-clear'/>
    <feature policy='require' name='stibp'/>
    <feature policy='require' name='arch-capabilities'/>
    <feature policy='require' name='xsaves'/>
    <feature policy='require' name='pdpe1gb'/>
    <feature policy='require' name='ibpb'/>
    <feature policy='require' name='amd-stibp'/>
    <feature policy='require' name='amd-ssbd'/>
    <feature policy='require' name='rdctl-no'/>
    <feature policy='require' name='ibrs-all'/>
    <feature policy='require' name='skip-l1dfl-vmentry'/>
    <feature policy='require' name='mds-no'/>
    <feature policy='require' name='pschange-mc-no'/>
    <feature policy='disable' name='hle'/>
    <feature policy='disable' name='rtm'/>
    <feature policy='disable' name='mpx'/>
    <feature policy='disable' name='intel-pt'/>
    <feature policy='disable' name='wbnoinvd'/>
  </cpu>
  <clock offset='utc'/>
  <on_poweroff>destroy</on_poweroff>
  <on_reboot>restart</on_reboot>
  <on_crash>destroy</on_crash>
  <devices>
    <emulator>/usr/bin/qemu-system-x86_64</emulator>
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/var/lib/libvirt/images/vagrant-libvirt-vagrant-libvirt_default.img' index='2'/>
      <backingStore type='file' index='3'>
        <format type='qcow2'/>
        <source file='/var/lib/libvirt/images/alvistack-VAGRANTSLASH-ubuntu-20.04_vagrant_box_image_20210425.0.0_0.img'/>
        <backingStore/>
      </backingStore>
      <target dev='vda' bus='virtio'/>
      <alias name='virtio-disk0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
    </disk>
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/var/lib/libvirt/images/vagrant-libvirt-vagrant-libvirt_default-vdb.qcow2' index='1'/>
      <backingStore/>
      <target dev='vdb' bus='virtio'/>
      <alias name='virtio-disk1'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x04' function='0x0'/>
    </disk>
    <controller type='usb' index='0' model='piix3-uhci'>
      <alias name='usb'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x2'/>
    </controller>
    <controller type='pci' index='0' model='pci-root'>
      <alias name='pci.0'/>
    </controller>
    <interface type='network'>
      <mac address='52:54:00:44:28:8f'/>
      <source network='vagrant-libvirt' portid='bb3069b0-92c0-47db-b19f-f7855db885f4' bridge='virbr1'/>
      <target dev='vnet0'/>
      <model type='virtio'/>
      <alias name='net0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x06' function='0x0'/>
    </interface>
    <serial type='pty'>
      <source path='/dev/pts/4'/>
      <target type='isa-serial' port='0'>
        <model name='isa-serial'/>
      </target>
      <alias name='serial0'/>
    </serial>
    <console type='pty' tty='/dev/pts/4'>
      <source path='/dev/pts/4'/>
      <target type='serial' port='0'/>
      <alias name='serial0'/>
    </console>
    <input type='mouse' bus='ps2'>
      <alias name='input0'/>
    </input>
    <input type='keyboard' bus='ps2'>
      <alias name='input1'/>
    </input>
    <graphics type='vnc' port='5900' autoport='yes' listen='127.0.0.1' keymap='en-us'>
      <listen type='address' address='127.0.0.1'/>
    </graphics>
    <video>
      <model type='cirrus' vram='16384' heads='1' primary='yes'/>
      <alias name='video0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
    </video>
    <memballoon model='virtio'>
      <stats period='5'/>
      <alias name='balloon0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x05' function='0x0'/>
    </memballoon>
  </devices>
  <seclabel type='dynamic' model='apparmor' relabel='yes'>
    <label>libvirt-778d5ab8-997e-4b11-82a6-cb6eb595cb32</label>
    <imagelabel>libvirt-778d5ab8-997e-4b11-82a6-cb6eb595cb32</imagelabel>
  </seclabel>
  <seclabel type='dynamic' model='dac' relabel='yes'>
    <label>+0:+0</label>
    <imagelabel>+0:+0</imagelabel>
  </seclabel>
</domain>
```

### PR DESCRIPTION
**Pull request by hswong3i**
Export Vagrant Box with `vagrant package` will trigger:

  - Shutdown VM if still running
  - Dump VM definination by `virsh dumpxml > box.xml`
  - Copy all attached disks for packaging

Therefore box structure updated as below:

    box.xml
    metadata.json
    Vagrantfile
    test-vda.qcow2
    test-vdb.qcow2

During `vagrant up` will trigger:

  - Read all box root disk filenames from `box.xml`
  - Copy box root disks from box archive to storage pool
  - Create snapshots for each root disks
  - Attach snapshots to running VM

Other changes:

  - Preserve both device vda/vdb/vdc/vdd for box root disks
  - Additional disks device starting from vde

Fixes #602
Fixes #1013
Fixes #1019
Fixes #1143
Fixes #1256

Signed-off-by: Wong Hoi Sing Edison <hswong3i@pantarei-design.com>

### Sample result of `vagrant package`:

```
$ ls -1d *
box.xml
metadata.json
Vagrantfile
test-vda.qcow2
test-vdb.qcow2

$ cat metadata.json
{
    "provider": "libvirt"
}

$ cat Vagrantfile
Vagrant.configure("2") do |config|
  config.vm.provider :libvirt do |libvirt|
    libvirt.driver = "kvm"
    libvirt.host = ""
    libvirt.connect_via_ssh = false
    libvirt.storage_pool_name = "default"
  end
end

user_vagrantfile = File.expand_path('../_include/Vagrantfile', __FILE__)
load user_vagrantfile if File.exists?(user_vagrantfile)

$ cat box.xml
<domain type='kvm'>
  <name>vagrant-libvirt-vagrant-libvirt_default</name>
  <uuid>da942062-9546-4eb8-95dc-3c21201f87b1</uuid>
  <memory unit='KiB'>524288</memory>
  <currentMemory unit='KiB'>524288</currentMemory>
  <vcpu placement='static'>1</vcpu>
  <os>
    <type arch='x86_64' machine='pc-i440fx-hirsute'>hvm</type>
    <boot dev='hd'/>
  </os>
  <features>
    <acpi/>
    <apic/>
    <pae/>
  </features>
  <cpu mode='host-model' check='partial'/>
  <clock offset='utc'/>
  <on_poweroff>destroy</on_poweroff>
  <on_reboot>restart</on_reboot>
  <on_crash>destroy</on_crash>
  <devices>
    <emulator>/usr/bin/qemu-system-x86_64</emulator>
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='vagrant-libvirt-vagrant-libvirt_default.qcow2'/>
      <target dev='vda' bus='virtio'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
    </disk>
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='vagrant-libvirt-vagrant-libvirt_default-vdb.qcow2'/>
      <target dev='vdb' bus='virtio'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x04' function='0x0'/>
    </disk>
    <controller type='usb' index='0' model='piix3-uhci'>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x2'/>
    </controller>
    <controller type='pci' index='0' model='pci-root'/>
    <interface type='network'>
      <mac address='52:54:00:16:02:b7'/>
      <source network='vagrant-libvirt'/>
      <model type='virtio'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x06' function='0x0'/>
    </interface>
    <serial type='pty'>
      <target type='isa-serial' port='0'>
        <model name='isa-serial'/>
      </target>
    </serial>
    <console type='pty'>
      <target type='serial' port='0'/>
    </console>
    <input type='mouse' bus='ps2'/>
    <input type='keyboard' bus='ps2'/>
    <graphics type='vnc' port='-1' autoport='yes' listen='127.0.0.1' keymap='en-us'>
      <listen type='address' address='127.0.0.1'/>
    </graphics>
    <video>
      <model type='cirrus' vram='16384' heads='1' primary='yes'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
    </video>
    <memballoon model='virtio'>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x05' function='0x0'/>
    </memballoon>
  </devices>
</domain>
```

### CODE REVIEW

**Comment by hswong3i** [comment]
Sample result of `vagrant package`:

```
$ ls -1d *
box.xml
metadata.json
Vagrantfile
vagrant-libvirt-vagrant-libvirt_default.qcow2
vagrant-libvirt-vagrant-libvirt_default-vdb.qcow2

$ cat metadata.json
{
    "provider": "libvirt"
}

$ cat Vagrantfile
Vagrant.configure("2") do |config|
  config.vm.provider :libvirt do |libvirt|
    libvirt.driver = "kvm"
    libvirt.host = ""
    libvirt.connect_via_ssh = false
    libvirt.storage_pool_name = "default"
  end
end

user_vagrantfile = File.expand_path('../_include/Vagrantfile', __FILE__)
load user_vagrantfile if File.exists?(user_vagrantfile)

$ cat box.xml
<domain type='kvm'>
  <name>vagrant-libvirt-vagrant-libvirt_default</name>
  <uuid>da942062-9546-4eb8-95dc-3c21201f87b1</uuid>
  <memory unit='KiB'>524288</memory>
  <currentMemory unit='KiB'>524288</currentMemory>
  <vcpu placement='static'>1</vcpu>
  <os>
    <type arch='x86_64' machine='pc-i440fx-hirsute'>hvm</type>
    <boot dev='hd'/>
  </os>
  <features>
    <acpi/>
    <apic/>
    <pae/>
  </features>
  <cpu mode='host-model' check='partial'/>
  <clock offset='utc'/>
  <on_poweroff>destroy</on_poweroff>
  <on_reboot>restart</on_reboot>
  <on_crash>destroy</on_crash>
  <devices>
    <emulator>/usr/bin/qemu-system-x86_64</emulator>
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='vagrant-libvirt-vagrant-libvirt_default.qcow2'/>
      <target dev='vda' bus='virtio'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
    </disk>
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='vagrant-libvirt-vagrant-libvirt_default-vdb.qcow2'/>
      <target dev='vdb' bus='virtio'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x04' function='0x0'/>
    </disk>
    <controller type='usb' index='0' model='piix3-uhci'>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x2'/>
    </controller>
    <controller type='pci' index='0' model='pci-root'/>
    <interface type='network'>
      <mac address='52:54:00:16:02:b7'/>
      <source network='vagrant-libvirt'/>
      <model type='virtio'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x06' function='0x0'/>
    </interface>
    <serial type='pty'>
      <target type='isa-serial' port='0'>
        <model name='isa-serial'/>
      </target>
    </serial>
    <console type='pty'>
      <target type='serial' port='0'/>
    </console>
    <input type='mouse' bus='ps2'/>
    <input type='keyboard' bus='ps2'/>
    <graphics type='vnc' port='-1' autoport='yes' listen='127.0.0.1' keymap='en-us'>
      <listen type='address' address='127.0.0.1'/>
    </graphics>
    <video>
      <model type='cirrus' vram='16384' heads='1' primary='yes'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
    </video>
    <memballoon model='virtio'>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x05' function='0x0'/>
    </memballoon>
  </devices>
</domain>
```

**Comment by hswong3i** [comment]
@electrofelix CI passed so looking forward for your kindly review. As minimal changes as possible.

Most diff are whitespace related please skip it for quick review: https://github.com/vagrant-libvirt/vagrant-libvirt/pull/1257/files?diff=unified&w=1

File: spec/unit/action/destroy_domain_spec.rb [hunk 1]
```diff
@@ -43,33 +43,33 @@
...
-          allow(libvirt_domain).to receive(:name).and_return('test')
-          allow(domain).to receive(:volumes).and_return([extra_disk], [root_disk])
-
-          expect(domain).to receive(:destroy).with(destroy_volumes: false)
-          expect(extra_disk).to receive(:destroy) # extra disk remove
-          expect(root_disk).to receive(:destroy)  # root disk remove
-          expect(subject.call(env)).to be_nil
-        end
-      end
+#      context 'when has additional disks' do
```
Comment (hswong3i) [inline spec/unit/action/destroy_domain_spec.rb:46]: @electrofelix Sorry that I couldn't manage this test case therefore temporarily disable it.

I tried with manually test by create Vagrantfile as below and it is functioning as expected ?_?

**Comment by electrofelix** [comment]
I don't have confidence that the XML outputted by libvirt is sufficiently portable between different versions to ensure that machines built on one distro can be reliably brought up on others. It doesn't seem that libvirt has provided the XML as a way of defining the entire VM and thus trying to use it as a box definition is setting up for having to work around every compatibility issue that is hidden depending on what version of libvirt and what extensions have been added by the distro.

Note to support domain XML as a box format, what is exported will need to be sanitized to remove distro specific as well as VM specific assignments to allow libvirt on the machine it is subsequently brought up on to regenerate the missing pieces. There are parts of the XML that are currently left undefined specifically to allow libvirt to make better decisions as well as recently noticed some of the network definition that we need to reduce how much is explicitly set in the current template to allow libvirt to intelligently replace the values.

There are multiple steps that need to be addressed before worrying about the code:
* Determine portability by having conversations with libvirt developers to understand whats expected and what pitfalls exist when using the XML generated for one version on a specific distro and attempting to use the outputted XML on significantly earlier versions (think using latest fedora to package and attempting to launch on one of the earlier RHEL that is still supported or the oldest ubuntu LTS).
* Talk to packer developers to understand the impact of switching to using libvirt XML for box images in the future (note they don't use libvirt to package boxes at the moment, instead packer talks directly to qemu)
* Look at other plugins potentially affected, how is vagrant-mutate supposed to convert between multi disk Virtualbox images to something understood by vagrant-libvirt.

The above all raise questions that have to be answered in detail before it's possible to switch to using the libvirt domain XML as part of the box definition.

As such, I intend to merge the multi disk support that uses the json as it is going to be the easiest for both packer and vagrant-mutate to adopt for the time being. I'm aware that there is still work to be done on the packaging end to make use of it, but one of the additional requirements will be for us to ensure the packaging can support both formats so end users can continue to package up using vagrant-libvirt while still using the oldest format until they are ready to switch.


If you want to look at what coding can be done before getting complete answers to the above questions from the different groups (libvirt, packer, and vagant-mutate devs), initial steps that can be looked at:
* Support callback hook for modifying the XML on create and start of the domain
   * In the event that there are issues with the XML provided in a box this will provide the required escape hatch to solving as needed. Additionally it will also provide the ability for people to modify the domain XML from vagrant-libvirt for scenarios not covered, so it's useful both now and in the event of migrating to using the domain XML for the box format.
* Ensuring additional disks are dynamically allocated a device after the box disks.
   * Simply reserving a number isn't going to be sufficient as one of the first use cases is to support ESX+VCSA running on top of kvm, where the VCSA typically would have 11 disks as part of it's deployment. Again this would solve the problem for both the multi disk support being added and lay the ground work for using the domain XML.

Any time we change the format being used, we will need to ensure that we're not making it difficult to work across versions of vagrant-libvirt. That means the packaging code needs a significant overhaul in order to support specifying packaging in any of the supported outputs where the VM can be exported to that format. Could say that V1 is the current, V2 will support multi disk, V3 might be a cleaned domain XML bundled. It will need to be possible after being able to package in V2 format for users to override and select V1 format instead where it's still a single disk output, or for the code to dynamically pick V1 where it's suitable. Where it's not possible and the user selects V1, the code will have to reject performing the packaging. The same will apply for V3, to have it as a default, it will be necessary support easy switching to the earlier formats where people need to support earlier versions of vagrant-libvirt with the box they are providing.

This also means that making the packaging format for the domain XML something requiring selection as the output is far more easier to accept as we can add warnings that the format is experimental and may not be adopted depending on the experience and impact to other tools working with vagrant-libvirt.

**Comment by hswong3i** [comment]
Yes you are correct: `dumpxml` is not target for portable but just for identical migration.

For portability we should looking for https://libguestfs.org/virt-v2v.1.html, especially https://libguestfs.org/virt-v2v-output-local.1.html

My solution is: don't export with `virt-v2v -i libvirt`, using `virt-v2v -i disk` instead (see https://libguestfs.org/virt-v2v.1.html#convert-disk-image-to-disk-image). The generated xml will therefore as minimal as possible for corresponding OS inside the provided disk image.

I am going to patch packer once PR accepted in this repo. I am using packer + vagrant-libvirt daily ;-)

In case `box.xml` not exists inside the box, we could simply provide a simple default for it internally? 

Sorry that is PR is too greedy and therefore lose the focus. #1264 should be much better to kick start with additional `box.xml` based packaging.

**Pull request closed without merging**